### DISCUSSION TITLE
How to create a Graphql query from OrchardCore Backend

### DISCUSSION THREAD

**Discussion by hyzx86** [OrchardCMS/OrchardCore#12097] [Q&A] [answered]
I hope I can use graphql query to export some data to Excel files, but this requires me to initiate a graphql query request in the background and then obtain the data. I found such a class in the source code

``` C# 
    public class OrchardGraphQLClient
    {
        public OrchardGraphQLClient(HttpClient client)
        {
            Client = client;
        }

        public ContentResource Content => new ContentResource(Client);

        public HttpClient Client { get; }
    }
```
As shown above, I need to build an httpclient before using it

At the same time, we need to fill all the information requested by the clientRequest into the new httpclient
This may require some tedious work. Has anyone implemented similar operations? 
Or do you have any better suggestions?

**Comment by hyzx86** [comment, marked as answer]
http:
``` C#
 private readonly HttpClient _httpClient;
    public ExcelController(IHttpClientFactory httpClientFactory)
    {
        _httpClient = httpClientFactory.CreateClient(Consts.GraphqlClient);
    }
    public async Task<ActionResult> ExportByGraphqlQuery(ExportByGraphqlQueryOptions options)
    {

#if DEBUG
        options.GraphQLQuery = @"  customer {    createdUtc
    custNum    dateeeee    datetimesssssss    displayText    guimo    modifiedUtc    name    publishedUtc    timeeeeeeeeeee  }";
#endif

        _httpClient.BaseAddress = new Uri($"{Request.Scheme}://{Request.Host}");
        _httpClient.DefaultRequestHeaders.Clear();
        if (!string.IsNullOrEmpty(Request.Headers["Cookie"]))
        {
            _httpClient.DefaultRequestHeaders.Add("Cookie", Request.Headers["Cookie"].ToString());
        }
        if (!string.IsNullOrEmpty(Request.Headers["Authorization"]))
        {
            _httpClient.DefaultRequestHeaders.Add("Authorization", Request.Headers["Authorization"].ToString());
        }

        var requestJson = new JObject(
        new JProperty("query", @"query { " + options.GraphQLQuery + " }")
        );
        if (options.QueryParams is not null)
        {
            requestJson["variables"] = JObject.FromObject(options.QueryParams);
        }
        var request = new HttpRequestMessage(HttpMethod.Post, "api/graphql");
        request.Content = new StringContent(requestJson.ToString(), Encoding.UTF8, "application/json");
        request.Headers.Accept.Add(new MediaTypeWithQualityHeaderValue("application/json"));
        var response = await _httpClient.SendAsync(request);

        if (!response.IsSuccessStatusCode)
        {
            throw new Exception(response.StatusCode + " " + await response.Content.ReadAsStringAsync());
        }

        var result = JObject.Parse(await response.Content.ReadAsStringAsync());
        return Json(result);//scuccess

    }
```

direct to excute graphql Query：

  https://github.com/OrchardCMS/OrchardCore/blob/c255456bf1d9093d13447e17485a23a1253f3e9b/src/OrchardCore.Modules/OrchardCore.Apis.GraphQL/GraphQLMiddleware.cs#L172-L189

**Reply by lampersky** [reply]
@hyzx86 if you are interested I can share later how I've achieved same thing but without http client

**Reply by hyzx86** [reply]
@lampersky 

I have used  `_executer.ExecuteAsync`  achieved the effect I needed,
But it would be great to see your code, Thanks! 😊

**Reply by lampersky** [reply]
@hyzx86 

I've created my own **GraphQLService**, which looks like this:

![image](https://user-images.githubusercontent.com/6403130/181165195-f0946446-6465-41ae-9b79-f5c08001793e.png)

and within my controller I'm able to perform queries directly from string or **GQL** files:

![image](https://user-images.githubusercontent.com/6403130/181164909-ef1afffc-22bd-4d72-8422-54ebc048c931.png)

_Please note, that it would be good to add some permission checks, which was omitted in my case._

**Reply by hyzx86** [reply]
Thanks for sharing

After reading your code, I am similar to your usage
In terms of permission, you should be able to directly use the permission setting of OC without additional inspection of the content
